README.md
# About

An SDK client for the RGB Software [OpenRGB](https://gitlab.com/CalcProgrammer1/OpenRGB/ "gitlab/CalcProgramer1/OpenRGB")

It is based on the already written, but apparently abandoned, [node.js OpenRGB](https://github.com/zebp/openrgb "github/vlakreeh/openrgb") library by [vlakreeh](https://github.com/zebp "vlakreeh") and the [Python implementation](https://github.com/jath03/openrgb-python "github/jath03/openrgb-python") by [jath03](https://github.com/jath03 "jath03").

It is still under heavy development, poorly written, and inconvenient to use, but since I'll probably forget this project in the next 5 seconds, I wanted to make this accessible for later generations of confused people.

# Installation

```
$ npm install openrgb-sdk
```

# Table of Content

- [About](#about)
- [Installation](#installation)
- [Table of Content](#table-of-content)
- [Guide](#guide)
	- [Setup](#setup)
		- [OpenRGB](#openrgb)
		- [Start the SDK-Server](#start-the-sdk-server)
		- [Node.js](#nodejs)
	- [Getting Started](#getting-started)
	- [Functions](#functions)
		- [Get device data](#get-device-data)
		- [Changing the LEDs of your device](#changing-the-leds-of-your-device)
			- [Setting it to a native mode](#setting-it-to-a-native-mode)
			- [Modify the LEDs yourself](#modify-the-leds-yourself)
		- [Make RGB objects](#make-rgb-objects)
- [Documentation](#documentation)
- [Examples](#examples)

# Guide

## Setup

### OpenRGB

Installation: [Gitlab OpenRGB Setup and Usage](https://gitlab.com/CalcProgrammer1/OpenRGB/-/wikis/OpenRGB-Windows-Setup-and-Usage)

### Start the SDK-Server

After successful installation click on the SDK-Client tab and click on `Start Server`, which should then set the Server Status to `online`. If you want it to start the SDK client on startup [follow this guide](https://www.howtogeek.com/228467/how-to-make-a-program-run-at-startup-on-any-computer/). Next, for Windows, right click on the icon and paste `--server` behind `target`, and for Linux paste `--server` behind `command`. This also applies if you are launching it through your console.

### Node.js

**most of the functions in openrgb-sdk rely on the javascript async/await functionality so you are going to have to put all of your code in a function similar to:**

```js
async function start () {
	
}

start()
```

## Getting Started

```js
const { Client } = require("openrgb-sdk")

async function start () {
	const client = new Client("Example", 6742, "localhost")
	
	await client.connect()

	/*
		- your code -
	*/

	await client.disconnect()
}

start()

```
You can familliarise yourself with this syntax [here.](https://javascript.info/async-await)

## Functions

### Get device data

You have two functions to get device data:

First, get the amount of devices connected to OpenRGB:

```js
	const amount = await client.getControllerCount()
```

Then, you can get the data of a specific device:

```js
	const device_0 = await client.getControllerData(0)
```

This function needs the ID of the device. The ID is just an ascending number, starting at 0, of the devices you have. They are given to the device just like they are listed in OpenRGB. This means, that if `getControllerCount()` gives back 2, you know that there are 3 devices: device 0, 1 and 2. With that knowledge you can just use a `for` loop to get all your devices.

```js
	let deviceList = []
	for (let deviceId = 0; deviceId < controllerCount; deviceId++) {
		deviceList.push(await client.getControllerData(deviceId))		
	}
```

### Changing the LEDs of your device

**[First of all ensure, that all other RGB programms on your PC are either uninstalled or closed. This includes background-programs.](https://gitlab.com/CalcProgrammer1/OpenRGB/-/wikis/OpenRGB-Windows-Setup-and-Usage#disable-other-rgb-applications)**

There are two ways to change the leds of your devices:

#### Setting it to a native mode

Every device has native modes you can set it to. They are saved on a chip inside of the device, and if you set a device to a mode, it will stay in this mode until it is changed by you or another RGB program in the background. This is more convenient, but almost impossible to synchronise across other devices. With the array you created before, you can now see the modes a device has and 
set a device to one of it's modes. This function takes the device id first, and then either accepts the name (as String/Text) or the id (as Number).

```js
	console.log(deviceList[0].modes) 
	await client.updateMode(0, deviceList[0].modes[5].id)
	await client.updateMode(1, deviceList[1].modes[3].name)
```

#### Modify the LEDs yourself

You can also change the LEDs of the device by yourself. These will also be set until changed, but you might want animated modes. They need the process to be running, and if its stopped the LEDs will be frozen in their color until they are changed. This is easier to sync, which is the why you'll probably use this more often. Also, you probably want to set the device before to the `direct` mode (which should be always mode 0), although they should be set to this at start.

The function first needs the device ID, and then an array of colors in RGB scheme. One item in the array resembles one LED, so to set all LEDs you have to create an array with the amount of LEDs the device has. If you only want one single color for the whole device you can just fill the array with the `.fill()` method.

```js
	const colors = Array(deviceList[2].colors.length).fill({
		red: 0xFF,
		green: 0xA5,
		blue: 0x00
	})
	
	console.log(`Setting the color of ${deviceList[2].name}`)
	await client.updateLeds(2, colors)
```

You can also set the LEDs for only one zone, or for a single LED, with simliar funtions.

An example for a custom mode using with updateLeds is found in `rainbow.js` in the `examples` folder.

### Make RGB objects

There are few ways to make RGB objects:

This is the regular way to of doing this, every value has to be set manually. All following ways output this kind of object:

```js
let violet = {
	red: 138,
	green: 43,
	blue: 226
}
```

This is quite long, and there are ways to compact it and/or use a different colour representation.

To simply compact it,

```js
const { utils } = require("openrgb-sdk")
let violet = utils.color(138, 43, 226)
```

To use a hex-string like CSS uses,

```js
const { utils } = require("openrgb-sdk")
let violet = utils.hexColor("#8a2be2")
```

to use the HSL-representation:

```js
const { utils } = require("openrgb-sdk")
let violet = utils.HSLColor(271.1, 0.759, 0.527)
```

or to use the HSV-representation:

```js
const { utils } = require("openrgb-sdk")
let violet = utils.HSVColor(271.1, 0.81, 0.886)
```

if any of the values are invalid or negative, they will default to 0. Values that are too high will be defaulted to their highest possible value.

# Documentation

A full documentation is available under [docs.m4rch.xyz](https://docs.m4rch.xyz/openrgb/).

# Examples

For examples, take a look at the `examples` folder in the downloaded package in your `node_modules` folder. 


CHANGELOG
version 0.5.0
[ general ]
	- dropped promise-socket dependency
[ functionality ]
	- deviceListUpdated event when openrgb detects a new device
	- getAllControllerData function which returns all devices
[ misc ]
	- command numbers added in utils
	- deleted useless tabs
	- many functions aren't async anymore
	- fixed jsdocs
	- failed connect doesn't throw error anymore ( bug, will hopefully be fixed in the future, for now use connect event)



src/client.js
const EventEmitter = require("events")
const { Socket } = require("net")
const bufferpack = require("bufferpack")

const utils = require("./utils.js")
const Device = require("./device")
const HEADER_SIZE = 16

module.exports = class Client extends EventEmitter {
	/**
	 * @param {string} name name for the client
	 * @param {number} port port of the connection
	 * @param {string} host host of the connection
	 */
	constructor (name, port, host) {
		super()

		this.name = name || "nodejs"
		this.port = +port || 6742
		this.host = host || "localhost"

		this.isConnected = false
		this.resolver = {}
	}
	/**
	 * connect to the OpenRGB-SDK-server
	 */
	connect () {
		this.socket = new Socket()
		this.socket.connect(this.port, this.host)

		this.isConnected = true

		this.socket.once("connect", () => {
			this.socket.on("close", () => {
				this.emit("disconnect")
				this.isConnected = false
			})
		})

		this.socket.on("readable", () => {
			let packet = this.socket.read()
			if (!packet) return
			if (this.resolver.resolve) {
				if (this.resolver.header) {
					if (this.resolver.header.length != packet.length) return //TODO return smth
					delete this.resolver.header
					return this.resolver.resolve(packet)
				}

				const headerBuffer = packet.slice(0, HEADER_SIZE)
				const header = this.decodeHeader(headerBuffer)

				if (packet.length < HEADER_SIZE + header.length) {
					return this.resolver.header = header
				}

				this.resolver.resolve(packet.slice(HEADER_SIZE))
				this.resolver = {}
			} else {
				if (packet.readUInt32LE(8) == utils.command.deviceListUpdated) {
					this.emit("deviceListUpdated")
				}
			}
		})

		this.socket.on("error", (err) => { 
			this.emit("error", err)
		})

		let nameBytes = new TextEncoder().encode(this.name)
		nameBytes = Buffer.concat([nameBytes, Buffer.from([0x00])])
		this.sendMessage(utils.command.setClientName, nameBytes)
		this.emit("connect")
	}
	/**
	 * disconnect from the OpenRGB-SDK-server
	 */
	disconnect () {
		if (this.isConnected) {
			this.socket.end()
			this.resolver = {}
		}
	}
	/**
	 * get the amount of devices
	 * @returns {Promise<number>}
	 */
	async getControllerCount () {
		this.sendMessage(utils.command.requestControllerCount)
		const buffer = await this.readMessage()
		return buffer.readUInt32LE()
	}
	/**
	 * get the protocol version
	 * @returns {Promise<number>}
	 */
	async getProtocolVersion () {
		this.sendMessage(utils.command.requestProtocolVersion)
		const buffer = await this.readMessage()
		return buffer.readUInt32LE()
	}
	/**
	 * get the properties of a device
	 * @param {number} deviceId the id of the requested device
	 * @returns {Promise<Device>}
	 */
	async getControllerData (deviceId) {
		this.sendMessage(utils.command.requestControllerData, undefined, deviceId)
		const buffer = await this.readMessage()
		return new Device(buffer, deviceId)
	}
	/**
	 * get the properties of all devices
	 * @returns {Promise<Device>[]}
	 */
	async getAllControllerData () {
		let devices = []
		for (let i = 0; i < await this.getControllerCount(); i++) {
			devices.push(await this.getControllerData(i))
		}
	}
	/**
	 * update all leds of a device
	 * @param {number} deviceId the id of the device
	 * @param {Object[]} colors the colors the device should be set to
	 */
	updateLeds (deviceId, colors) {
		const size = 2 + (4 * colors.length)

		const colorsBuffer = Buffer.alloc(size)
		colorsBuffer.writeUInt16LE(colors.length)

		for (let i = 0; i < colors.length; i++) {
			const color = colors[i]
			const offset = 2 + (i * 4)
			colorsBuffer.writeUInt8(color.red, offset)
			colorsBuffer.writeUInt8(color.green, offset + 1)
			colorsBuffer.writeUInt8(color.blue, offset + 2)
		}

		const prefixBuffer = Buffer.alloc(4)
		prefixBuffer.writeUInt32LE(size)

		this.sendMessage(utils.command.updateLeds, Buffer.concat([prefixBuffer, colorsBuffer]), deviceId)
	}
	/**
	 * update all leds of a device
	 * @param {number} deviceId the id of the device
	 * @param {number} zoneId the id of the zone
	 * @param {Object[]} colors the colors the zone should be set to
	 */
	updateZoneLeds (deviceId, zoneId, colors) {
		const size = 6 + (4 * colors.length)
		const colorsBuffer = Buffer.alloc(size)
		colorsBuffer.writeUInt32LE(zoneId)
		colorsBuffer.writeUInt16LE(colors.length, 4)
		for (let i = 0; i < colors.length; i++) {
			const color = colors[i]
			const offset = 6 + (i * 4)
			colorsBuffer.writeUInt8(color.red, offset)
			colorsBuffer.writeUInt8(color.green, offset + 1)
			colorsBuffer.writeUInt8(color.blue, offset + 2)
		}
		const prefixBuffer = Buffer.alloc(4)
		prefixBuffer.writeUInt32LE(size)
		this.sendMessage(utils.command.updateZoneLeds, Buffer.concat([prefixBuffer, colorsBuffer]), deviceId)
	}
	/**
	 * update all leds of a device
	 * @param {number} deviceId the id of the device
	 * @param {number} ledId the id of the led
	 * @param {Object} colors the color the led should be set to
	 */
	updateSingleLed (deviceId, ledId, color) {
		let buff = Buffer.concat([bufferpack.pack("<I", [ledId]), bufferpack.pack("<BBB", [color.red, color.green, color.blue]), new Buffer.alloc(1)])
		this.sendMessage(utils.command.updateSingleLed, buff, deviceId)
	}
	/**
	 * sets the device to its mode for individual led control
	 * @param {number} deviceId the id of the requested device
	 * @returns {Promise<Device>}
	 */
	setCustomMode (deviceId) {
		this.sendMessage(utils.command.setCustomMode, undefined, deviceId)
	}
	/**
	 * update all leds of a device
	 * @param {number} deviceId the id of the device
	 * @param {number|string} mode either the id or name of a mode
	 * @param {Object} custom the settings of this mode e.g. speed, color etc.
	 */
	async updateMode (deviceId, mode, custom) {
		//TODO: shorten and beautify
		let modeData

		if (typeof deviceId != "number") throw new Error("arg deviceId not given")
		let device = await this.getControllerData(deviceId)

		if (typeof mode == "number") {
			if (!device.modes.filter(elem => elem.id == mode)[0]) throw new Error("Number given is not the id of a mode")
			modeData = device.modes.filter(elem => elem.id == mode)[0]
		} else if (typeof mode == "string") {
			if (!device.modes.filter(elem => elem.name.toLowerCase() == mode.toLowerCase())[0]) throw new Error("Name given is not the name of a mode")
			modeData = device.modes.filter(elem => elem.name == mode)[0]
		} else throw new Error("arg mode is either not given or not an instance of string or number")

		if (typeof custom != "object" && typeof custom != "undefined") throw new Error("3rd argument must be either an object ore undefined")

		if (typeof custom != "undefined") {
			if (custom.random) {
				if (!modeData.flagList.includes("randomColor")) throw new Error("Random color can't be chosen")
				modeData.colorMode = 3
			} else if (modeData.flagList.includes("modeSpecificColor")) {
				modeData.colorMode = 2
			} else if (modeData.flagList.includes("perLedColor")) {
				modeData.colorMode = 1
			} else {
				modeData.colorMode = 0
			}
			if (!isNaN(+custom.speed)) {
				if (modeData.flagList.includes("speed")) {
					if (custom.speed < modeData.speedMin || custom.speed > modeData.speedMax) throw new Error("Speed either to high or to low")
					if ((custom.speed % 1) != 0) throw new Error("Speed must be a round number")
					modeData.speed = custom.speed
				} else throw new Error("Speed can't be set for this mode")
			}
			if (!isNaN(+custom.direction)) {
				if (modeData.flagList.includes("direction")) {
					let arr = ["directionLR", "directionUD", "directionHV"]
					if (modeData.flagList.includes(arr[Math.floor(custom.direction / 2)])) {
						modeData.direction = custom.direction
					} else throw new Error("Can't set direction to this value")

					if ((custom.direction % 1) != 0) throw new Error("Direction must be a round number")

				} else throw new Error("Direction can't be set for this mode")
			}
			if (custom.colors) {
				if (modeData.colorMode == 1) {
					this.updateLeds(deviceId, custom.colors)
				} else {
					if (custom.colors.length > modeData.colorMax) throw new Error("Too many colors.")
					if (modeData.colorLength <= 0) throw new Error("Color can't be set for this mode")
					modeData.colors = custom.colors
				}
			}
		}

		let data = Buffer.concat([
			bufferpack.pack("<I", [modeData.id]),
			this.pack_string(modeData.name),
			bufferpack.pack("<IIIIIIIII", [
				modeData.value, 
				modeData.flags, 
				modeData.speedMin, 
				modeData.speedMax, 
				modeData.colorMin, 
				modeData.colorMax, 
				modeData.speed, 
				modeData.direction, 
				modeData.colorMode
			]),
			this.pack_list(modeData.colors ? modeData.colors : 0), 
		])
		data = Buffer.concat([bufferpack.pack("<I", [data.length, bufferpack.calcLength("<I")]), data])

		this.sendMessage(utils.command.updateMode, data, deviceId)
	}
	/**
	 * update all leds of a device
	 * @param {number} deviceId the id of the device
	 * @param {number} zoneId the id of the zone
	 * @param {number} zoneLength the length the zone should be set to
	 */
	resizeZone (deviceId, zoneId, zoneLength) {
		this.sendMessage(utils.command.resizeZone, bufferpack.pack("<ii", [zoneId, zoneLength]), deviceId)
	}
	/**
	 * @private
	 */
	sendMessage (commandId, buffer = Buffer.alloc(0), deviceId) {
		if (!this.isConnected) throw new Error("can't write to socket if not connected to OpenRGB")
		const header = this.encodeHeader(commandId, buffer.byteLength, deviceId)
		const packet = Buffer.concat([header, buffer])
		this.socket.write(packet)
	}
	/**
	 * @private
	 */
	async readMessage () {
		if (!this.isConnected) throw new Error("can't read from socket if not connected to OpenRGB")
		return new Promise(resolve => this.resolver.resolve = resolve)
	}
	/**
	 * @private
	 */
	encodeHeader (commandId, length, deviceId) {
		const buffer = Buffer.alloc(HEADER_SIZE)

		let index = buffer.write("ORGB", "ascii")
		index = buffer.writeUInt32LE(deviceId, index)
		index = buffer.writeUInt32LE(commandId, index)
		index = buffer.writeUInt32LE(length, index)

		return buffer
	}
	/**
	 * @private
	 */
	decodeHeader (buffer) {
		const deviceId = buffer.readUInt32LE(4)
		const commandId = buffer.readUInt32LE(8)
		const length = buffer.readUInt32LE(12)
		return { deviceId, commandId, length }
	}
	/**
	 * @private
	 */
	pack_list (arr) {
		let out = bufferpack.pack("<H", [arr.length])
		arr.forEach(element => {
			out = Buffer.concat([out, new Buffer.from(""), bufferpack.pack("<BBBx", [element.red, element.green, element.blue])])
		})
		return out
	}
	/**
	 * @private
	 */
	pack_string (string) {
		return Buffer.concat([bufferpack.pack(`<H${string.length}s`, [string.length + 1, encodeURIComponent(string)]), Buffer.from('\x00')])
	}
}
